internal/command/apply_test.go

```diff
@@ -1000,56 +976,223 @@ func TestApply_planVarsEphemeral_applyTime(t *testing.T) {
+func TestApply_planVarsEphemeral_applyTime(t *testing.T) {
+	for name, tc := range map[string]func(*testing.T, *ApplyCommand, string, string, func(*testing.T) *terminal.TestOutput){
+		// Test first that an apply supplying only the apply-time variable "foo"
+		// succeeds.
+		"only passing ephemeral variable": func(t *testing.T, c *ApplyCommand, statePath, planPath string, done func(*testing.T) *terminal.TestOutput) {
+			args := []string{
+				"-state", statePath,
+				"-var", "foo=bar",
+				planPath,
+			}
+			code := c.Run(args)
+			output := done(t)
+			if code != 0 {
+				t.Fatal("should've succeeded: ", output.Stderr())
+			}
 		},
-	}
-	args = []string{
-		"-state", statePath,
-		planPath,
-	}
-	code = c.Run(args)
-	output = done(t)
-	if code == 0 {
-		t.Fatal("should've failed: ", output.Stdout())
+
+		// Now test that supplying "bar", which is not an apply-time variable, fails.
+		"passing non-ephemeral variable": func(t *testing.T, c *ApplyCommand, statePath, planPath string, done func(*testing.T) *terminal.TestOutput) {
+			args := []string{
+				"-state", statePath,
+				"-var", "foo=bar",
+				"-var", "bar=bar",
+				planPath,
+			}
+			code := c.Run(args)
+			output := done(t)
+			if code == 0 {
+				t.Fatal("should've failed: ", output.Stdout())
+			}
+		},
+
+		// Test that the apply also fails if we do *not* supply a value for
+		// the apply-time variable foo.
+		"missing ephemeral variable": func(t *testing.T, c *ApplyCommand, statePath, planPath string, done func(*testing.T) *terminal.TestOutput) {
+			args := []string{
+				"-state", statePath,
+				planPath,
+			}
+			code := c.Run(args)
+			output := done(t)
+			if code == 0 {
+				t.Fatal("should've failed: ", output.Stdout())
+			}
+		},
+
+		"passing ephemeral variable through vars file": func(t *testing.T, c *ApplyCommand, statePath, planPath string, done func(*testing.T) *terminal.TestOutput) {
+			const planVarFile = `
+foo = "bar"
+`
+
+			// Write a tfvars file with the variable
+			tfVarsPath := testVarsFile(t)
+			err := os.WriteFile(tfVarsPath, []byte(planVarFile), 0600)
+			if err != nil {
+				t.Fatalf("Could not write vars file %e", err)
+			}
+
+			args := []string{
+				"-state", statePath,
+				"-var-file", tfVarsPath,
+				planPath,
+			}
+			code := c.Run(args)
+			output := done(t)
+			if code != 0 {
+				t.Fatal("should've succeeded: ", output.Stderr())
+			}
+		},
+
+		"passing ephemeral variable through environment variable": func(t *testing.T, c *ApplyCommand, statePath, planPath string, done func(*testing.T) *terminal.TestOutput) {
+			t.Setenv("TF_VAR_foo", "bar")
+
+			args := []string{
+				"-state", statePath,
+				planPath,
+			}
+			code := c.Run(args)
+			output := done(t)
+			if code != 0 {
+				t.Fatal("should've succeeded: ", output.Stderr())
+			}
+		},
+
+		// "passing ephemeral variable through interactive prompts": func(t *testing.T, c *ApplyCommand, statePath, planPath string, done func(*testing.T) *terminal.TestOutput) {
```

Comment:
Is this something we intend to support?

Reply:
I think we do, I was planning on doing it in a separate PR though.

Reply:
hmm, I just tried the interactive prompt workflow, which is broken by this. Is there an easy way to disable interactive prompts in plan for now? The most common case for interactive prompts is someone simply running a combined plan+apply with `terraform apply`, which fails with `No value for required variable`.

Reply:
I fixed this now by re-using the ephemeral input from plan during apply if we have a combined plan and apply workflow, the apply with a plan fails without prompting for interactive input, are we ok with this?
It seemed like the most straight-forward solution to me.